[문자열 덧셈 계산기] 김동혁 미션 제출합니다.

BaSak0630 opened the pull request.

BaSak0630 committed:
docs : Add function list
기능 구현 이전 구현해야할 기능 목록을 정리 후 추가합니다.

BaSak0630 committed:
feat : Add print start message
시작 문구 출력 기능을 추가했습니다.

BaSak0630 committed:
docs : Edit function list
시작 문구 출력 기능 추가로 구현 todo 리스트 내용 변경

BaSak0630 committed:
feat : Add read String
사용자로 부터 문자열 입력 받기 기능 추가

BaSak0630 committed:
docs : Edit function list
사용자로 부터 문자열 입력 받기 기능 추가로 구현 todo 리스트 내용 변경

BaSak0630 committed:
feat : Add split function
사용자로 부터 받은 문자열 숫자를 구분자로 나누고 문자를 숫자로 변경 기능 추가

BaSak0630 committed:
docs : Edit function list
구분자를 이용해 숫자 나눈고 문자를 int로 변환기능 구현 후 todo 리스트 내용 변경

BaSak0630 committed:
feat : Add custom separator
입력 받은 문자열에서 커스텀 구분자를 등록 기능 추가

BaSak0630 committed:
docs : Edit function list
커스텀 구분자 추가 기능 구현으로 todo  리스트 현황 변경

BaSak0630 committed:
refactor : Refactoring
별도의 변수로 인스턴시에이션 해서 변경했습니다.

BaSak0630 committed:
feat : Add get sum numbers
입력받은 숫자들의 합산을 구하는 기능 추가

BaSak0630 committed:
docs : Edit function list
숫자합산 기능 추가후 list 변경

BaSak0630 committed:
feat : Add print result
합산 결과를 출력하는 기능 추가

BaSak0630 committed:
docs : Edit function list
결과 출력기능 추가 후 list 변경

BaSak0630 committed:
fix : Fix Special character exception
“|”, “*”, “^” 등의 특수문자는  \를 붙여야 split() 함수의 인자로 적용가능합니다,. 그래서 특수문자를 일반 문자 , 로 치환해서 해결했습니다.

BaSak0630 committed:
refactor : Edit code
불필요하게 된 변수를 지우고 OutputView에 Numbers객체 레퍼런스를 넘겨주는 식으로 처리했습니다.

BaSak0630 committed:
refactor : Edit code
getNumberList에서 바로 Numbers 객체를 돌려주는 식을 수정했습니다.

BaSak0630 committed:
refactor : Edit function name
기능에 맞는 함수 이름으로 변경

BaSak0630 committed:
refactor : Refactoring add addNumberToNumberList()
numberArray에 요소를 int로 변환 후 list 추가하는 메서드로 변경

BaSak0630 committed:
fix : input string is empty check
빈 공백을 입력 받는 경우 예외 처리되게 수정

BaSak0630 committed:
fix : Change method split
기존에 커스텀 구분자의 문자열을 요소로 나누어 여러 캐릭터들을 전부구분자로 사용하는 방식에서 전체문자열을 묶어서 한개의 구분자로
자르게 변경하였습니다.

BaSak0630 committed:
refactor : Refactoring variable name
특정 문자열을 static 데이터 멤버로 변경

BaSak0630 committed:
docs : Edit function list
예외 처리 완료 후 todo lsit 내용변경

BaSak0630 committed:
fix : Change integer to Double
양수를 생각했을떄 처음에 단순히 양의 부호인 수를 생각했습니다. 하지만 소수도 양수에 포함 되기 떄문에 기존에 integer를 실수인 Double로 변경

BaSak0630 committed:
test : Add printResult test
OutputView의 printResult() 메서드에 Numbers객체에 합의 소수점 여부에 따라 출력 테스트 추가

BaSak0630 committed:
test : Add NumbersTest
Numbers 객체 음수 검증과 Numbers의 ArrayList의 합산결과 검증

BaSak0630 committed:
test : Add Separator test
Separator test 추가

BaSak0630 committed:
test : Edit Separator test
Separator test 수정

BaSak0630 committed:
refactor : Change access modifier
접근제어자 변경

BaSak0630 committed:
test : Edit Separator test
Separator test 수정

BaSak0630 committed:
test : Add InputViewTest

BaSak0630 committed:
refactor : Change class name
calculatorController를 calculator로 변경했습니다. 의미적으로 calculator가 더 맞는것 같아 서 변경

BaSak0630 committed:
refactor : Edit indentation
들여쓰기 수정

bowook commented on src/main/java/calculator/view/OutputView.java:
```diff
@@ -0,0 +1,23 @@
+    public void printResult(Numbers numbers) {
+        Double sum = numbers.getSum();
+        if(sum%1 == 0) {
+            String result = String.format("%.0f", sum);
```
코드에 대한 분석을 보니, 양의 실수를 처리하기 위해 Double 타입을 선택한 점이 인상적입니다.
특히 문제의 요구사항을 충실히 반영하려는 노력이 보이는 코드인 것 같아요!
Double 타입을 사용할 경우 근삿값이 나올 수 있다는 점에서 %.0f 형식을 활용하여 이를 처리하려고 한 시도가 눈에 띄네요!

다만, Double 타입은 오차가 발생할 수 있는 특성이 있기 때문에 정확한 계산이 필요한 경우에는 BigDecimal 타입을 사용하는 방법을 고려해 볼 수 있습니다. 이를 통해 오차를 최소화할 수 있는 장점이 있으니, 해당 기능도 알아보면 좋겠습니다!

BaSak0630 replied:
BigDecimal 타입은 생각도 안해보았습니다. 존재 자체는 알고 있는데 사용해본적이 없어서 익숙하지 않습니다. 부동 소수점에 대한 내용을 조금만 알고 있어서 조금 더 공부해보고 나중에 코드에 적용해보겠습니다. 감사합니다.

bowook replied:
답변해주셔서 감사합니다!
1주차 고생하셨습니다!

bowook commented on src/test/java/calculator/NumbersTest.java:
```diff
@@ -0,0 +1,37 @@
+    }
+    @Test
+    @DisplayName("Numbers 덧셈 검증")
+    void getSum() throws Exception{
```
테스트 코드에서 given, when, then 구조를 통해 테스트를 짠 부분이 매우 인상적입니다.

BaSak0630 replied:
감사합니다. ㅎㅎ

bowook reviewed (commented):
1주 차 고생하셨습니다!
요구사항을 상세히 분석하시려고 노력하신 흔적이 보이는 것 같아요!

BaSak0630 committed:
refactor : Add final

다른 개발자가 이 코드를 봤을때 final이 붙어 있으면 필드 변경에
좀 더 신중할 수 있을거 같습니다!
불변임을 명시하여 다른사람과 같이 프로그래밍 할때 의도적으로 변경의 여지를 없는 것이 가능할것 같습니다.

gyuoo commented on src/main/java/calculator/Numbers.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+
+public class Numbers {
+    private List<Double> numbers;
```
Double로 실수까지 처리하셨군요! 저는 양의 정수만 생각하고 실수 처리는 전혀 생각지 못했습니다..
디테일 좋습니다!

sunkong25 replied:
저도요! 정수만 생각하고 실수 처리는 못했는데 하나 더 배워갑니다:)

gyuoo commented on src/main/java/calculator/Numbers.java:
```diff
@@ -0,0 +1,28 @@
+            sum += number;
+        }
+        return sum;
+    }
```
overflow에 대한 고려가 되어있지 않은 것으로 확인되는데, 맞을까요?

BaSak0630 replied:
overflow에 대해 간과했던것 같습니다. 알려주셔서 감사합니다.

gyuoo commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,77 @@
+        return numberList;
+    }
+
+    private List<Double> addNumberToNumberList(String[] numberArray) {
+        List<Double> numberList = new ArrayList<>();
+        for(String s : numberArray){
+            if(!s.isEmpty()){
+                Double i = Double.parseDouble(s);
+                numberList.add(i);
+            }
+        }
+        return numberList;
+    }
```
공통 피드백에 Collection 지향에 대한 얘기가 있던 것으로 기억하는데, 배열의 split도 이용하며 Collection까지 고려하신 것이 좋네요!

gyuoo commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,77 @@
+        return numberList;
+    }
+
+    private String getNumberString(String readString) {
+        int subStringInt = readString.indexOf(CUSTOM_SEPARATOR_END) + CUSTOM_SEPARATOR_END.length();
+        String substring = readString.substring(subStringInt);
+        return substring;
+    }
```
커스텀 구분자가 중간에 있는 경우도 고려된 것 같네요!
`readString.indexOf(CUSTOM_SEPARATOR_END)`가 -1인 경우나
`\\n1:2:3`과 같은 입력도 예외 처리를 했다면 더 좋았을 것 같습니다.

BaSak0630 replied:
getNumberString메서드가 호출되기 이전에 hasCustomSeparator 이라는 메서드를 통해서 readString.indexOf(CUSTOM_SEPARATOR_END) 가 -1인 경우의 예외처리를 하고 있습니다. 
마찬가지로 \\n1:2:3 입력의 경우에도 hasCustomSeparator 에서  예외처리를 하고 있습니다.

gyuoo commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    public boolean hasCustomSeparator(String readString) {
+        if(readString.contains(CUSTOM_SEPARATOR_START) && readString.contains(CUSTOM_SEPARATOR_END)){
+            return true;
+        }
+        return false;
```
`return readString.contains(CUSTOM_SEPARATOR_START) && readString.contains(CUSTOM_SEPARATOR_END)` 은 어떤가요?

BaSak0630 replied:
제가 놓친것 같습니다. 알려주셔서 감사합니다. ㅎㅎ

gyuoo commented on src/main/java/calculator/view/InputView.java:
```diff
@@ -0,0 +1,13 @@
+        String s = Console.readLine();
+        if(s.isEmpty()) {
+            throw new IllegalArgumentException("[ERROR]덧셈할 문자열을 입력해 주세요");
+        }
```
요구사항에 빈 값이 입력되는 경우 0을 출력해야 하는 것으로 알고 있습니다!

BaSak0630 replied:
알려주셔서 감사합니다. 이걸 놓쳤었네요 ㅠㅠ

gyuoo commented on src/test/java/calculator/NumbersTest.java:
```diff
@@ -0,0 +1,37 @@
+    @Test
+    @DisplayName("Numbers 덧셈 검증")
+    void getSum() throws Exception{
+        //given
+        List<Double> numberList = new ArrayList<>();
+        numberList.add(1.0);
+        numberList.add(2.0);
+        numberList.add(3.0);
+        Numbers numbers = new Numbers(numberList);
+        //when
+        Double sum = numbers.getSum();
+        //then
+        assertEquals(6.0, sum);
```
given / when / then 구조를 이용하셨군요! 저도 이번 주에는 적용해보겠습니다

gyuoo reviewed (commented):
리뷰 요청 주셔서 감사드립니다.
잘 읽었고 제 리뷰가 조금이라도 도움이 되길 바랍니다.
이번 주도 힘내세요!

sunkong25 commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,77 @@
+    private static final String CUSTOM_SEPARATOR_START = "//";
+    private static final String CUSTOM_SEPARATOR_CENTER = "|";
+    private static final String CUSTOM_SEPARATOR_CENTER_REGEX = "\\|";
+    private static final String CUSTOM_SEPARATOR_END = "\\n";
```
저는 일부만 상수로 사용했는데 문자열을 모두 상수로 사용하는 것이 좋은 것 같아요!
의미가 담겨있어서 어디에 사용하는지 알 수 있고, 쓸 때마다 헷갈리지 않고 사용할 수 있을 것 같아요:)

sunkong25 commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,77 @@
+
+    private List<Double> split(String numberString) {
+        List<Double> numberList;
+        String[] symbolArray = symbols.split(CUSTOM_SEPARATOR_CENTER_REGEX);
```
symbols로 문자열을 합쳤다가 symbolArray로 나눠서 다시 저장하신 것 같은데 왜 이런 방법을 사용했는지 궁금합니다!

BaSak0630 replied:
구분자를 추가할 때 특수 문자나 숫자 등 문제가 발생할수 있는 문자들을 전부 ','로 치환해서 split을 해주면 문제가 발생하지 않을 것 같아서 다시 배열에 저장하였습니다.

sunkong25 commented on src/main/java/calculator/view/OutputView.java:
```diff
@@ -0,0 +1,23 @@
+    private static final String START_MESSAGE = "덧셈할 문자열을 입력해 주세요.";
+
+    public void printStartMessage() {
+        System.out.println(START_MESSAGE);
```
문자열은 상수 처리해서 사용한 것이 인상깊었습니다:)

sunkong25 reviewed (commented):
다른 분들께서 리뷰를 자세히 달아주셔서 저는 궁금한 점을 달아봤습니다!
좋은 코드 덕분에 많이 배워갑니다. 
1주차 수고 많으셨습니다! 2주차도 같이 파이팅해요☺️

BaSak0630 committed:
fix : Fix input empty

요구사항에 빈 값이 입력되는 경우 0을 출력해야 합니다.

BaSak0630 committed:
refactor : Edit return

지역변수와 if 줄이기
